File: app/src/main/java/com/laushkina/anastasia/psychosophy/view/relationships/RelationshipsViewModel.java
package com.laushkina.anastasia.psychosophy.view.relationships;

import android.databinding.BaseObservable;
import android.databinding.Bindable;
import android.opengl.Visibility;
import android.text.Html;
import android.view.View;

import com.laushkina.anastasia.psychosophy.BR;

public class RelationshipsViewModel extends BaseObservable {
    private CharSequence firstFunctionRelationshipsTitle;
    private CharSequence firstFunctionRelationships;
    private CharSequence secondFunctionRelationshipsTitle;
    private CharSequence secondFunctionRelationships;
    private CharSequence thirdFunctionRelationshipsTitle;
    private CharSequence thirdFunctionRelationships;
    private CharSequence forthFunctionRelationshipsTitle;
    private CharSequence forthFunctionRelationships;

    private int imageAndHintVisibility;

    RelationshipsViewModel(){
        firstFunctionRelationshipsTitle = "";
        firstFunctionRelationships = "";
        secondFunctionRelationshipsTitle = "";
        secondFunctionRelationships = "";
        thirdFunctionRelationshipsTitle = "";
        thirdFunctionRelationships = "";
        forthFunctionRelationshipsTitle = "";
        forthFunctionRelationships = "";
        imageAndHintVisibility = View.VISIBLE;
    }

    //TODO builder?
    RelationshipsViewModel(String firstTitle, String firstDescription,
                           String secondTitle, String secondDescription,
                           String thirdTitle, String thirdDescription,
                           String forthTitle, String forthDescription,
                           int imageAndHintVisibility){
        setFirstFunctionRelationshipsTitle(firstTitle);
        setFirstFunctionRelationships(firstDescription);
        setSecondFunctionRelationshipsTitle(secondTitle);
        setSecondFunctionRelationships(secondDescription);
        setThirdFunctionRelationshipsTitle(thirdTitle);
        setThirdFunctionRelationships(thirdDescription);
        setForthFunctionRelationshipsTitle(forthTitle);
        setForthFunctionRelationships(forthDescription);
        setImageAndHintVisibility(imageAndHintVisibility);
    }

    public void refreshWith(RelationshipsViewModel newValue){
        setFirstFunctionRelationshipsTitle(newValue.firstFunctionRelationshipsTitle);
        setFirstFunctionRelationships(newValue.firstFunctionRelationships);
        setSecondFunctionRelationshipsTitle(newValue.secondFunctionRelationshipsTitle);
        setSecondFunctionRelationships(newValue.secondFunctionRelationships);
        setThirdFunctionRelationshipsTitle(newValue.thirdFunctionRelationshipsTitle);
        setThirdFunctionRelationships(newValue.thirdFunctionRelationships);
        setForthFunctionRelationshipsTitle(newValue.forthFunctionRelationshipsTitle);
        setForthFunctionRelationships(newValue.forthFunctionRelationships);
    }

    @Bindable
    public CharSequence getFirstFunctionRelationships() {
        return firstFunctionRelationships;
    }

    private void setFirstFunctionRelationships(String firstFunctionRelationships) {
        setFirstFunctionRelationships(Html.fromHtml(firstFunctionRelationships));
    }

    private void setFirstFunctionRelationships(CharSequence firstFunctionRelationships) {
        this.firstFunctionRelationships = firstFunctionRelationships;
        notifyPropertyChanged(BR.firstFunctionRelationships);
    }

    @Bindable
    public CharSequence getSecondFunctionRelationships() {
        return secondFunctionRelationships;
    }

    private void setSecondFunctionRelationships(String secondFunctionRelationships) {
        setSecondFunctionRelationships(Html.fromHtml(secondFunctionRelationships));
    }

    private void setSecondFunctionRelationships(CharSequence secondFunctionRelationships) {
        this.secondFunctionRelationships = secondFunctionRelationships;
        notifyPropertyChanged(BR.secondFunctionRelationships);
    }

    @Bindable
    public CharSequence getThirdFunctionRelationships() {
        return thirdFunctionRelationships;
    }

    private void setThirdFunctionRelationships(String thirdFunctionRelationships) {
        setThirdFunctionRelationships(Html.fromHtml(thirdFunctionRelationships));
    }

    private void setThirdFunctionRelationships(CharSequence thirdFunctionRelationships) {
        this.thirdFunctionRelationships = thirdFunctionRelationships;
        notifyPropertyChanged(BR.thirdFunctionRelationships);
    }

    @Bindable
    public CharSequence getForthFunctionRelationships() {
        return forthFunctionRelationships;
    }

    private void setForthFunctionRelationships(String forthFunctionRelationships) {
        setForthFunctionRelationships(Html.fromHtml(forthFunctionRelationships));
    }

    private void setForthFunctionRelationships(CharSequence forthFunctionRelationships) {
        this.forthFunctionRelationships = forthFunctionRelationships;
        notifyPropertyChanged(BR.forthFunctionRelationships);
    }

    @Bindable
    public CharSequence getFirstFunctionRelationshipsTitle() {
        return firstFunctionRelationshipsTitle;
    }

    private void setFirstFunctionRelationshipsTitle(String firstFunctionRelationshipsTitle) {
        setFirstFunctionRelationshipsTitle(Html.fromHtml(firstFunctionRelationshipsTitle));
    }

    private void setFirstFunctionRelationshipsTitle(CharSequence firstFunctionRelationshipsTitle) {
        this.firstFunctionRelationshipsTitle = firstFunctionRelationshipsTitle;
        notifyPropertyChanged(BR.firstFunctionRelationshipsTitle);
    }

    @Bindable
    public CharSequence getSecondFunctionRelationshipsTitle() {
        return secondFunctionRelationshipsTitle;
    }

    private void setSecondFunctionRelationshipsTitle(String secondFunctionRelationshipsTitle) {
        setSecondFunctionRelationshipsTitle(Html.fromHtml(secondFunctionRelationshipsTitle));
    }

    private void setSecondFunctionRelationshipsTitle(CharSequence secondFunctionRelationshipsTitle) {
        this.secondFunctionRelationshipsTitle = secondFunctionRelationshipsTitle;
        notifyPropertyChanged(BR.secondFunctionRelationshipsTitle);
    }

    @Bindable
    public CharSequence getThirdFunctionRelationshipsTitle() {
        return thirdFunctionRelationshipsTitle;
    }

    private void setThirdFunctionRelationshipsTitle(String thirdFunctionRelationshipsTitle) {
        setThirdFunctionRelationshipsTitle(Html.fromHtml(thirdFunctionRelationshipsTitle));
    }

    private void setThirdFunctionRelationshipsTitle(CharSequence thirdFunctionRelationshipsTitle) {
        this.thirdFunctionRelationshipsTitle = thirdFunctionRelationshipsTitle;
        notifyPropertyChanged(BR.thirdFunctionRelationshipsTitle);
    }

    @Bindable
    public CharSequence getForthFunctionRelationshipsTitle() {
        return forthFunctionRelationshipsTitle;
    }

    private void setForthFunctionRelationshipsTitle(String forthFunctionRelationshipsTitle) {
        setForthFunctionRelationshipsTitle(Html.fromHtml(forthFunctionRelationshipsTitle));
    }

    private void setForthFunctionRelationshipsTitle(CharSequence forthFunctionRelationshipsTitle) {
        this.forthFunctionRelationshipsTitle = forthFunctionRelationshipsTitle;
        notifyPropertyChanged(BR.forthFunctionRelationshipsTitle);
    }

    @Bindable
    public int getImageAndHintVisibility() {
        return imageAndHintVisibility;
    }

    public void setImageAndHintVisibility(int imageAndHintVisibility) {
        this.imageAndHintVisibility = imageAndHintVisibility;
        notifyPropertyChanged(BR.imageAndHintVisibility);
    }
}

File: app/src/main/java/com/laushkina/anastasia/psychosophy/view/relationships/RelationshipsViewModel.kt
package com.laushkina.anastasia.psychosophy.view.relationships

import androidx.databinding.BaseObservable
import androidx.databinding.Bindable
import android.text.Html
import android.view.View
import com.laushkina.anastasia.psychosophy.BR

class RelationshipsViewModel(private var firstFunctionRelationshipsTitle: CharSequence?,
                             private var firstFunctionRelationships: CharSequence?,
                             private var secondFunctionRelationshipsTitle: CharSequence?,
                             private var secondFunctionRelationships: CharSequence?,
                             private var thirdFunctionRelationshipsTitle: CharSequence?,
                             private var thirdFunctionRelationships: CharSequence?,
                             private var forthFunctionRelationshipsTitle: CharSequence?,
                             private var forthFunctionRelationships: CharSequence?,
                             private var imageAndHintVisibility: Int) : BaseObservable() {
    constructor() : this("", "", "", "", "", "", "", "", View.VISIBLE)

    constructor(firstTitle: String,
                firstDescription: String,
                secondTitle: String,
                secondDescription: String,
                thirdTitle: String,
                thirdDescription: String,
                forthTitle: String,
                forthDescription: String,
                imageAndHintVisibility: String) : this()

    init {
        setFirstFunctionRelationshipsTitle(firstFunctionRelationshipsTitle)
        setFirstFunctionRelationships(firstFunctionRelationships)
        setSecondFunctionRelationshipsTitle(secondFunctionRelationshipsTitle)
        setSecondFunctionRelationships(secondFunctionRelationships)
        setThirdFunctionRelationshipsTitle(thirdFunctionRelationshipsTitle)
        setThirdFunctionRelationships(thirdFunctionRelationships)
        setForthFunctionRelationshipsTitle(forthFunctionRelationshipsTitle)
        setForthFunctionRelationships(forthFunctionRelationships)
        setImageAndHintVisibility(imageAndHintVisibility)
    }

    fun refreshWith(newValue: RelationshipsViewModel) {
        setFirstFunctionRelationshipsTitle(newValue.firstFunctionRelationshipsTitle)
        setFirstFunctionRelationships(newValue.firstFunctionRelationships)
        setSecondFunctionRelationshipsTitle(newValue.secondFunctionRelationshipsTitle)
        setSecondFunctionRelationships(newValue.secondFunctionRelationships)
        setThirdFunctionRelationshipsTitle(newValue.thirdFunctionRelationshipsTitle)
        setThirdFunctionRelationships(newValue.thirdFunctionRelationships)
        setForthFunctionRelationshipsTitle(newValue.forthFunctionRelationshipsTitle)
        setForthFunctionRelationships(newValue.forthFunctionRelationships)
    }

    @Bindable
    fun getFirstFunctionRelationships(): CharSequence? {
        return firstFunctionRelationships
    }

    private fun setFirstFunctionRelationships(firstFunctionRelationships: CharSequence?) {
        this.firstFunctionRelationships = fromHml(firstFunctionRelationships)
        notifyPropertyChanged(BR.firstFunctionRelationships)
    }

    @Bindable
    fun getSecondFunctionRelationships(): CharSequence? {
        return secondFunctionRelationships
    }

    private fun setSecondFunctionRelationships(secondFunctionRelationships: CharSequence?) {
        this.secondFunctionRelationships = fromHml(secondFunctionRelationships)
        notifyPropertyChanged(BR.secondFunctionRelationships)
    }

    @Bindable
    fun getThirdFunctionRelationships(): CharSequence? {
        return thirdFunctionRelationships
    }

    private fun setThirdFunctionRelationships(thirdFunctionRelationships: CharSequence?) {
        this.thirdFunctionRelationships = fromHml(thirdFunctionRelationships)
        notifyPropertyChanged(BR.thirdFunctionRelationships)
    }

    @Bindable
    fun getForthFunctionRelationships(): CharSequence? {
        return forthFunctionRelationships
    }

    private fun setForthFunctionRelationships(forthFunctionRelationships: CharSequence?) {
        this.forthFunctionRelationships = fromHml(forthFunctionRelationships)
        notifyPropertyChanged(BR.forthFunctionRelationships)
    }

    @Bindable
    fun getFirstFunctionRelationshipsTitle(): CharSequence? {
        return firstFunctionRelationshipsTitle
    }

    private fun setFirstFunctionRelationshipsTitle(firstFunctionRelationshipsTitle: CharSequence?) {
        this.firstFunctionRelationshipsTitle = firstFunctionRelationshipsTitle
        notifyPropertyChanged(BR.firstFunctionRelationshipsTitle)
    }

    @Bindable
    fun getSecondFunctionRelationshipsTitle(): CharSequence? {
        return secondFunctionRelationshipsTitle
    }

    private fun setSecondFunctionRelationshipsTitle(secondFunctionRelationshipsTitle: CharSequence?) {
        this.secondFunctionRelationshipsTitle = secondFunctionRelationshipsTitle
        notifyPropertyChanged(BR.secondFunctionRelationshipsTitle)
    }

    @Bindable
    fun getThirdFunctionRelationshipsTitle(): CharSequence? {
        return thirdFunctionRelationshipsTitle
    }

    private fun setThirdFunctionRelationshipsTitle(thirdFunctionRelationshipsTitle: CharSequence?) {
        this.thirdFunctionRelationshipsTitle = thirdFunctionRelationshipsTitle
        notifyPropertyChanged(BR.thirdFunctionRelationshipsTitle)
    }

    @Bindable
    fun getForthFunctionRelationshipsTitle(): CharSequence? {
        return forthFunctionRelationshipsTitle
    }

    private fun setForthFunctionRelationshipsTitle(forthFunctionRelationshipsTitle: CharSequence?) {
        this.forthFunctionRelationshipsTitle = forthFunctionRelationshipsTitle
        notifyPropertyChanged(BR.forthFunctionRelationshipsTitle)
    }

    @Bindable
    fun getImageAndHintVisibility(): Int {
        return imageAndHintVisibility
    }

    fun setImageAndHintVisibility(imageAndHintVisibility: Int) {
        this.imageAndHintVisibility = imageAndHintVisibility
        notifyPropertyChanged(BR.imageAndHintVisibility)
    }

    private fun fromHml(charSequence: CharSequence?): CharSequence {
        if (charSequence != null && charSequence is String) {
            return Html.fromHtml(charSequence)
        }
        return charSequence ?: ""
    }
}